### PR TITLE
Move browser safe proxy logic into an authorizer

### LINKED ISSUE #16619: pod proxying not working
This used to work before:
```
curl --http1.1 -kv -H"Authorization: Bearer XXX" https://cluster:8443/api/v1/namespaces/ns/pods/podX1:8080/proxy/appEndPoint
```
But not anymore on Openshift 3.7. I get the error:
```
User "jwendell" cannot get pods/unsafeproxy in the namespace "ns"
```
Did anything in this regard changed in 3.7? Does the user need some special permission now?

##### Version
oc v3.7.0-alpha.0+e5b4a1c-454
kubernetes v1.7.0+695f48a16f
features: Basic-Auth

Server xxx
openshift v3.7.0-0.126.4
kubernetes v1.7.0+80709908fd

### LINKED ISSUE #16710: Proxy action always resolves as unsafeproxy, even with authenticated users
Requests to the proxy subresource are transformed to be the `unsafeproxy` action, even when the users are authenticated and not coming from a direct browser request.  

The comment at https://github.com/openshift/origin/blob/050c9eaa0d758e66f576b97da9aef1d9559a563c/pkg/authorization/authorizer/browser_safe_request_info_resolver.go#L56 seems to imply that RBAC subresource should be `proxy` for authenticated users / users not coming directly from a browser, and `unsafeproxy` for non-authenticated users or direct browser requests.

However, according to https://github.com/openshift/origin/blob/050c9eaa0d758e66f576b97da9aef1d9559a563c/pkg/cmd/server/kubernetes/master/master_config.go#L403 and above, the filter is installed in RequestInfoResolver, which runs *before* the generic WithAuthentication filter: https://github.com/openshift/origin/blob/050c9eaa0d758e66f576b97da9aef1d9559a563c/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/config.go#L491.  This ordering indicates that the browser-safe proxy filter will *never* get user information from the context (the context will always be missing the appropriate key for the user, so `UserFrom` returns that the information is missing).

This breaks the HPA, whose policy expects to be able to `/proxy` the Heapster service.  We could change the policy, but it seems like this filter just isn't working as intended.

##### Version
master as of 050c9eaa0d758e66f576b97da9aef1d9559a563c

##### Steps To Reproduce
1. Fire up OpenShift
2. Attempt a curl (with certs or a bearer token) to `/api/v1/namespaces/openshift-infra/services/https:heapster:/proxy/apis` or any other use of the proxy subresource, and notice that we receive an error that our user cannot `unsafeproxy`, not that the user can't `proxy`.

Debugged this with @frobware 
cc @deads2k since you seem to have been the last one to shuffle this code around ;-)

### PR DESCRIPTION
**Pull request by enj**
This change moves the browser safe proxy verification from a request info resolver to an authorizer.  It is required because we moved to the upstream filters in 689cdeefd85752b2af78433fb99c8b11fc11f428.  The upstream handler chain resolves request info, then authentication and then authorization.  Our old chain would mutate request info after authentication and before authorization.  This meant that we had access to the user info data.  Since the upstream handler chain does any mutation to the request info before authentication, our browser safe proxy request info resolver would never see any user data.  This meant that it would incorrectly identify safe proxy requests as unsafe.  To continue using the upstream chain, we must move this logic into an authorizer that wraps our existing authorizers.  This will guarantee that it has access to the authentication information it requires.  Since the authorizer only has access to authorization attributes, it cannot make any decisions based on the context or request info.  In this case, it means that we can no longer honor the X-CSRF-Token header as a way of confirming that the request is not originating from a browser.  This is an acceptable break from legacy behavior as that header was never documented as a way of passing the browser check.  The most likely user of this feature is the proxy command, which is easy to use in an authenticated manner with either oc or kubectl.  Administrators can continue to grant the unsafe verb to unauthenticated users should they desire to bypass this safety check entirely.

Signed-off-by: Monis Khan <mkhan@redhat.com>

Fixes #16619
Fixes #16710

/assign @liggitt @deads2k

@openshift/sig-security

cc @jwendell @DirectXMan12 since you guys encountered this issue.

### CODE REVIEW

File: pkg/authorization/authorizer/browsersafe/authorizer.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	if user := attributes.GetUser(); user != nil {
+		if a.authenticatedGroups.HasAny(user.GetGroups()...) {
+			// An authenticated request indicates this isn't a browser page load.
+			// Browsers cannot make direct authenticated requests.
+			// This depends on the API not enabling basic or cookie-based auth.
+			return attributes
+		}
+	}
+
+	browserSafeAttributes := authorizer.AttributesRecord{
```
Comment (liggitt) [inline pkg/authorization/authorizer/browsersafe/authorizer.go]: This is actually a good use case for inlining, and just implementing GetVerb and GetSubresource. We don't want to have to update this if new attributes are added

File: pkg/cmd/server/origin/authorizer.go [hunk 2]
```diff
@@ -56,15 +57,18 @@ func newAuthorizer(
...
+	// Wrap with an authorizer that detects unsafe requests and modifies verbs/resources appropriately so policy can address them separately
+	browserSafeAuthorizer := browsersafe.NewBrowserSafeAuthorizer(scopeLimitedAuthorizer, bootstrappolicy.AuthenticatedGroup)
 
 	graph := node.NewGraph()
 	node.AddGraphEventHandlers(graph, podInformer, pvInformer)
 	nodeAuthorizer := node.NewAuthorizer(graph, nodeidentifier.NewDefaultNodeIdentifier(), kbootstrappolicy.NodeRules())
 
 	authorizer := authorizerunion.New(
-		authorizerfactory.NewPrivilegedGroups(user.SystemPrivilegedGroup), // authorizes system:masters to do anything, just like upstream
+		authorizerfactory.NewPrivilegedGroups(bootstrappolicy.MastersGroup), // authorizes system:masters to do anything, just like upstream
```
Comment (liggitt) [inline pkg/cmd/server/origin/authorizer.go]: Why not use the kube constant?

File: test/integration/authorization_test.go [hunk 3]
```diff
@@ -1550,8 +1550,7 @@ func TestOldLocalResourceAccessReviewEndpoint(t *testing.T) {
 	}
 }
 
-// TestClusterPolicyCache confirms that the creation of cluster role bindings fallback to live lookups when the referenced cluster role is not cached
-func TestClusterPolicyCache(t *testing.T) {
```
Comment (liggitt) [inline test/integration/authorization_test.go]: Why removed? Flaky?

Comment (enj) [reply test/integration/authorization_test.go]: This was supposed to be removed as part of one of the earlier RBAC clean ups.  There is no policy, and no cache to test anymore.

File: pkg/cmd/server/origin/authorizer.go (hunk 2 above)
Comment (enj) [reply pkg/cmd/server/origin/authorizer.go]: Allowed me to drop an import.

Comment (liggitt) [reply pkg/cmd/server/origin/authorizer.go]: I'd rather stick with the upstream constant

File: pkg/cmd/server/origin/authorizer.go [hunk 4]
```diff
@@ -56,15 +57,18 @@ func newAuthorizer(
 	subjectLocator := authorizer.NewSubjectLocator(kubeSubjectLocator)
 
 	scopeLimitedAuthorizer := scope.NewAuthorizer(roleBasedAuthorizer, clusterRoleGetter, messageMaker)
+	// Wrap with an authorizer that detects unsafe requests and modifies verbs/resources appropriately so policy can address them separately
+	browserSafeAuthorizer := browsersafe.NewBrowserSafeAuthorizer(scopeLimitedAuthorizer, bootstrappolicy.AuthenticatedGroup)
```
Comment (liggitt) [inline pkg/cmd/server/origin/authorizer.go]: also, can use `user.AllAuthenticated` from upstream package

File: test/integration/authorization_test.go [hunk 5]
```diff
@@ -1562,23 +1561,43 @@ func TestClusterPolicyCache(t *testing.T) {
...
+	}{
+		{
+			name:          "safe to proxy",
+			client:        userClient.Core().RESTClient(),
+			expectedError: `User "user" cannot get pods/proxy in the namespace "ns": User "user" cannot get pods/proxy in project "ns"`,
+		},
+		{
+			name:          "unsafe to proxy",
+			client:        anonymousClient,
+			expectedError: `User "system:anonymous" cannot get pods/proxy in the namespace "ns": User "system:anonymous" cannot get pods/unsafeproxy in project "ns"`,
```
Comment (liggitt) [inline test/integration/authorization_test.go]: hardcoding the full error will be fragile... check for `pods/unsafeproxy` in one, and absence of `pods/unsafeproxy` in the other?

Comment (liggitt) [reply test/integration/authorization_test.go]: also check leading `proxy` verb?

File: pkg/authorization/authorizer/browsersafe/authorizer.go (hunk 1 above)
Comment (deads2k) [reply pkg/authorization/authorizer/browsersafe/authorizer.go]: > This is actually a good use case for inlining, and just implementing GetVerb and GetSubresource. We don't want to have to update this if new attributes are added

This is still outstanding.

**Comment by deads2k** [comment]
lgtm after jordan's comments

**Comment by enj** [comment]
/kind bug

Added LGTM since comments addressed.

File: pkg/cmd/server/kubernetes/master/master_config.go [hunk 6]
```diff
@@ -76,6 +75,9 @@ import (
 	openapigenerated "github.com/openshift/origin/pkg/openapi"
 	securityapi "github.com/openshift/origin/pkg/security/apis/security"
 	"github.com/openshift/origin/pkg/version"
+
+	// TODO fix this install, it is required for TestPreferredGroupVersions to pass
+	_ "github.com/openshift/origin/pkg/authorization/apis/authorization/install"
```
Comment (enj) [inline pkg/cmd/server/kubernetes/master/master_config.go]: I added this install import per @deads2k recommendation to make `TestPreferredGroupVersions` pass.

Re-adding LGTM.

**Comment by enj** [comment]
Flake #16716

**Pull request merged**